src/Database/Relations/BelongsToMany.php

```diff
@@ -100,6 +100,110 @@ public function create(array $attributes = [], array $pivotData = [], $sessionKe
         return $model;
     }
 
+    /**
+     * @param mixed $id
+     * @param array $attributes
+     * @param bool  $touch
+     */
+    public function attach($id, array $attributes = [], $touch = true)
+    {
+        $insertData = $this->formatAttachRecords($this->parseIds($id), $attributes);
+        $elementIDList = array_pluck($insertData, $this->relatedPivotKey);
+
+        /**
+         * @event model.relation.beforeAttach
+         * Called before creating a new relation between models (only for BelongsToMany relation)
+         *
+         * Example usage:
+         *
+         *     $model->bindEvent('model.relation.beforeAttach', function (string $relationName, array $elementIDList, array $insertData) use (\October\Rain\Database\Model $model) {
+         *         if (!$model->isRelationValid($elementIDList)) {
+         *             throw new \Exception("Invalid relation!");
+         *             return false;
+         *         }
+         *     });
+         *
+         */
+        $eventDataList = $this->parent->fireEvent('model.relation.beforeAttach', [$this->relationName, $elementIDList, $insertData]);
+        if (!empty($eventDataList)) {
```

Comment:
No need for this $eventDataList handling code, the `$halt` parameter should be passed to fireEvent instead simplifying to 
```php
if ($this->parent->fireEvent('model.relation.beforeAttach', [$this->relationName, $elementIDList, $insertData], true) {
    return;
}
```

See http://octobercms.com/docs/api/october/rain/support/traits/emitter#fireevent